### PR TITLE
Generalize evolution actions to arbitrary temporal_ids

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This PR adds a `step_prefix` template alias to the tags that may be used as a `temporal_id`:

- `Tags::TimeId` uses `Tags::dt`
- `LinearSolver::Tags::IterationId` uses `LinearSolver::Tags::OperatorAppliedTo`

This `step_prefix` is used by `ComputeTimeDerivative` (for evolution systems, renamed from `ComputeVolumeDuDt`), `ComputeOperatorAction` (for elliptic systems) and `ApplyFluxes` (for both, renamed from `ApplyBoundaryFluxesGlobalTimestepping`) to prefix the `variables_tag` for storing the results of their computations.

I also started to add documentation to actions on what types they require of the `Metavariables` and `system`. The types are explained in more detail in the `Evolution/Elliptic systems` Doxygen groups. I suggest we keep improving on this documentation in the future.

**What breaks**:
- `Actions::ComputeVolumeDuDt ` must be renamed to `Actions::ComputeTimeDerivative`
- `dg::Actions::ApplyBoundaryFluxesGlobalTimestepping ` must be renamed to `dg::Actions::ApplyFluxes`
- `system::du_dt` must be renamed to `system::compute_time_derivative`.

### Types of changes:

- [x] New feature

### Component:

- [x] Code
- [x] Documentation

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by nilsvu** [comment]
It's probably best to look at the commits one-by-one. The last commit touches a number of files, but it only propagates the changes from the first commits to the codebase.

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp [hunk 1]
```diff
@@ -31,57 +31,69 @@ struct Mesh;
...
                     const ArrayIndex& /*array_index*/,
                     const ActionList /*meta*/,
                     const ParallelComponent* const /*meta*/) noexcept {
-    static_assert(not Metavariables::local_time_stepping,
-                  "ApplyBoundaryFluxesGlobalTimeStepping cannot be used with "
-                  "local time-stepping.");
-
     using system = typename Metavariables::system;
-    constexpr size_t volume_dim = system::volume_dim;
+    constexpr size_t Dim = system::volume_dim;
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp]: we use `snake_case` for variable names, `dim` would be preferred :)

File: src/Evolution/Executables/Burgers/EvolveBurgers.hpp [hunk 2]
```diff
@@ -9,7 +9,6 @@
 #include "Domain/DomainCreators/RegisterDerivedWithCharm.cpp"
```
Comment (nilsdeppe) [inline src/Evolution/Executables/Burgers/EvolveBurgers.hpp:9]: I think this commit needs to be squashed with the second once review is done so each individual commit compiles, right?

**Review by nilsdeppe — COMMENTED** [review]
Just one minor stylistic comment that I'm happy for you to squash.

@wthrowe @kidder could you please look at this when you get a chance? Thanks!

File: src/DataStructures/DataBox/DataBoxTag.hpp [hunk 3]
```diff
@@ -598,6 +598,30 @@ struct variables_tag_with_tags_list_impl<DispatchTagType::Prefix> {
   template <typename Tag, typename NewTagsList>
   using f = typename helper<Tag, NewTagsList>::type;
 };
+
+// Implementation of get_variables_tags_list_impl
```
Comment (wthrowe) [inline src/DataStructures/DataBox/DataBoxTag.hpp]: Remove "_impl".

File: docs/GroupDefs.hpp [hunk 4]
```diff
@@ -397,8 +397,27 @@
...
+ * fluxes for. Use `db::add_tag_prefix<LinearSolver::Tags::Operand, fields_tag>`
+ * unless you have a reason not to.
+ * - `volume`: A struct that computes the volume contribution to the DG
+ * operator. Must expose a `tmpl::list` of `argument_tags` and a static `apply`
+ * function that takes the following arguments in this order:
+ *   - The types of the tensors in
+ * `db::add_tag_prefix<Metavariables::temporal_id::step_prefix, variables_tag>`
+ * (which represent the linear operator applied to the variables) as not-null
+ * pointers.
+ *   - The types of the `argument_tags` as constant references.
```
Comment (wthrowe) [inline docs/GroupDefs.hpp]: How are these last two supposed to be exposed by the system?  And what are `argument_tags` the arguments to?

File: docs/GroupDefs.hpp [hunk 5]
```diff
@@ -418,7 +437,30 @@
...
+ * - `variables_tag`: The evolved variables to compute DG volume contributions
+ * and fluxes for.
+ * - `volume`: A struct that computes the volume contribution to the DG
+ * discretization of the time derivative. Must expose a `tmpl::list` of
+ * `argument_tags` and a static `apply` function that takes the following
+ * arguments in this order:
+ *   - The types of the tensors in
+ * `db::add_tag_prefix<Metavariables::temporal_id::step_prefix, variables_tag>`
+ * (which represent the time derivatives of the variables) as not-null pointers.
+ *   - The types of the `argument_tags` as constant references.
```
Comment (wthrowe) [inline docs/GroupDefs.hpp:456]: Same comments as above.

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp (hunk 1 above)
Comment (wthrowe) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp]: I don't see a point in changing it from the original `volume_dim`.

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+ * variables_tag>`, where `step_prefix` is retrieved from the
+ * `Metavariables::temporal_id`. For instance, in a time evolution this prefix
+ * would represent a time derivative `Tags::dt`.
+ *
+ * Uses:
+ * - Metavariables:
+ *   - `temporal_id::step_prefix`
+ * - System:
+ *   - `variables_tag`
+ *   - `volume`
```
Comment (wthrowe) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp]: I don't like the name "volume" here.  There are a lot of things associated with the volume, this should indicate which one it is.  (In fact, this action doesn't know anything about DG, so the "volume" association isn't really right anyway, but SpECTRE is sufficiently DG-focused that I think that's OK.)

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+ * - DataBox:
+ *   - db::add_tag_prefix<step_prefix, variables_tag>
+ *   - All elements in `volume::argument_tags`
+ *
+ * DataBox changes:
+ * - Modifies:
+ *   - `db::add_tag_prefix<step_prefix, variables_tag>`
+ */
+template <typename Metavariables>
+struct ComputeVolume {
```
Comment (wthrowe) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp]: Same comment as for the system tag.  We're not computing the volume itself, but the volume terms or similar.

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp [hunk 8]
```diff
@@ -31,57 +31,69 @@ struct Mesh;
...
+ * DataBox changes:
+ * - Modifies:
+ *   - `db::add_tag_prefix<step_prefix, variables_tag>`
+ *   - `flux_comm_types::mortar_data_tag`
+ */
+template <typename Metavariables>
+struct ApplyFluxes {
+  template <typename Tag>
+  using step_prefix =
+      typename Metavariables::temporal_id::template step_prefix<Tag>;
```
Comment (wthrowe) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp]: This is only used once.  If you inline it you can avoid templating the action.

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  template <typename Tag>
+  using step_prefix =
+      typename Metavariables::temporal_id::template step_prefix<Tag>;
```
Comment (wthrowe) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp]: This is only used once.  It you inline it you can avoid templating the action.

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/CMakeLists.txt [hunk 10]
```diff
@@ -3,9 +3,10 @@
 
 set(LIBRARY_SOURCES
   ${LIBRARY_SOURCES}
-  Actions/Test_ApplyBoundaryFluxesGlobalTimeStepping.cpp
   Actions/Test_ApplyBoundaryFluxesLocalTimeStepping.cpp
+  Actions/Test_ApplyFluxes.cpp
   Actions/Test_ComputeNonconservativeBoundaryFluxes.cpp
+  Actions/Test_ComputeVolume.cpp
```
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/CMakeLists.txt]: This isn't DG-specific.  Don't move it into DG.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/DataStructures/DataBox/DataBoxTag.hpp [hunk 11]
```diff
@@ -598,6 +598,30 @@ struct variables_tag_with_tags_list_impl<DispatchTagType::Prefix> {
...
+template <DispatchTagType TagType>
+struct get_variables_tags_list_impl;
+}  // namespace DataBox_detail
+
+template <typename Tag>
+using get_variables_tags_list =
```
Comment (wthrowe) [inline src/DataStructures/DataBox/DataBoxTag.hpp:608]: Needs a test.

File: docs/GroupDefs.hpp (hunk 4 above)
Comment (nilsvu) [reply docs/GroupDefs.hpp]: This refers to the volume computer struct, the list is indented

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp (hunk 8 above)
Comment (nilsvu) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp]: The compiler complained when I tried that so I had to go with the type alias, do you know how I can make it work inline?

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp (hunk 1 above)
Comment (nilsvu) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp]: agreed

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp (hunk 6 above)
Comment (nilsvu) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp]: Agreed, but I couldn’t think of a better name. Do you have a suggestion?

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/CMakeLists.txt (hunk 10 above)
Comment (nilsvu) [reply tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/CMakeLists.txt]: Not sure what the distinction is between dg and non-dg. If this is a volume contribution to the dg discretization of the System it may as well be in dg. Where else should I move it (must be outside of evolution)?

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp (hunk 6 above)
Comment (nilsvu) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp]: Actually, everything that I can think of seems redundant (volume terms/contribution/operator/...). If what we mean with „system“ are the PDEs in some discretization, then this operator does indeed describe the volume part of it, so `system::volume` seems appropriate. Thoughts?

Comment (nilsleiffischer) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp]: Actually, everything that I can think of seems redundant (volume terms/contribution/operator/...). If what we mean with „system“ are the PDEs in some discretization, then this operator does indeed describe the volume part of it, so `system::volume` seems appropriate. Thoughts?

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp (hunk 7 above)
Comment (nilsvu) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp]: If we keep this in the `dg` namespace then this means the volume contribution to the dg discretisation. Only if we try to be too general with this does it become ambiguous I think. But if we really want to generalize we may as well go all the way and have an `Action::MutateApply` that works with any struct that has an appropriate interface for `db::mutate_apply`. So my proposal with this PR is to have this DG-specialized `dg::Actions::ComputeVolume` that retrieves what it needs from Metavariables/System. For more general applications we should add `Action:MutateApply`.

File: docs/GroupDefs.hpp (hunk 4 above)
Comment (wthrowe) [reply docs/GroupDefs.hpp]: Oops, I missed that.

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp (hunk 8 above)
Comment (wthrowe) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp]: `using dt_variables_tag = db::add_tag_prefix<Metavariables::temporal_id::template step_prefix, variables_tag>`

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp (hunk 6 above)
Comment (wthrowe) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp]: I agree that the "system" is the (discretized) PDEs, but PDEs don't have a volume or a volume part.  And actually I've reconsidered what I said above due to the discussion on moving this class and I don't want "volume" in this name at all.

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/CMakeLists.txt (hunk 10 above)
Comment (wthrowe) [reply tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/CMakeLists.txt]: This action was designed to be usable for any ODE system, not just those derived from DG or even from PDEs.  We have intended non-PDE use cases for this action.  (Various ray-tracing problems, mostly.)

This is actually already done in the tests.  In moving this you made the "Time" tests depend on the `dg` namespace, which they should not need.

The "Volume" in the name was kind of inappropriate, and I'd actually intended to drop it whenever I (or someone else) added the first non-DG system.

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp (hunk 7 above)
Comment (wthrowe) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp]: A general action might be the right way to go here, yeah.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp (hunk 7 above)
Comment (kidder) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp]: Yes, @nilsdeppe and I have also discussed introducing an `Action::MutateApply` as this pattern occurs frequently.  If it allows us to avoid an ambiguous name in this case, maybe it is worth introducing now.

**Comment by nilsvu** [comment]
Refactored in response to our discussion. We now have `ComputeTimeDerivative` for evolution systems, `ComputeOperatorAction` for elliptic systems and `ApplyFluxes` for both.

**Comment by nilsvu** [comment]
@wthrowe @nilsdeppe @kidder this should be ready

File: tests/Unit/DataStructures/DataBox/Test_DataBoxTag.cpp [hunk 12]
```diff
@@ -140,6 +140,12 @@ static_assert(
     "Failed testing variables_tag_with_tags_list");
 /// [variables_tag_with_tags_list]
 
+static_assert(
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/DataBox/Test_DataBoxTag.cpp:143]: A test with 2 tags would be a good addition (though unlikely to be different, 1 is a fair special case) :)

File: docs/GroupDefs.hpp [hunk 13]
```diff
@@ -397,8 +397,27 @@
  * \details Actions and parallel components may require an elliptic system to
```
Comment (nilsdeppe) [inline docs/GroupDefs.hpp:397]: In the commit message `allows elliptic code` -> `allows the elliptic code`

**Review by nilsdeppe — COMMENTED** [review]
A couple minor things, I'm happy for you to squash :)

File: src/Evolution/Actions/ComputeTimeDerivative.hpp [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+template <typename Metavariables>
+class ConstGlobalCache;
+}  // namespace Parallel
+/// \endcond
+
+namespace Actions {
+/*!
+ * \ingroup ActionsGroup
+ * \ingroup DiscontinuousGalerkinGroup
+ * \brief Compute the bulk contribution to the time derivative of the variables
```
Comment (wthrowe) [inline src/Evolution/Actions/ComputeTimeDerivative.hpp]: Reword to not assume DG.

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp [hunk 15]
```diff
@@ -59,27 +74,23 @@ struct ApplyBoundaryFluxesGlobalTimeStepping {
...
                     const ParallelComponent* const /*meta*/) noexcept {
-    static_assert(not Metavariables::local_time_stepping,
-                  "ApplyBoundaryFluxesGlobalTimeStepping cannot be used with "
-                  "local time-stepping.");
-
     using system = typename Metavariables::system;
     constexpr size_t volume_dim = system::volume_dim;
     using variables_tag = typename system::variables_tag;
-    using dt_variables_tag = db::add_tag_prefix<Tags::dt, variables_tag>;
+    using dt_variables_tag =
```
Comment (wthrowe) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp:80]: Do you want to rename this?  (Up to you.)

File: tests/Unit/Evolution/Actions/Test_ComputeTimeDerivative.cpp [hunk 16]
```diff
@@ -21,12 +21,17 @@
...
+  using step_prefix = Tags::dt<Tag>;
+};
+
 struct var_tag : db::SimpleTag {
   using type = int;
   static std::string name() noexcept { return "var_tag"; }
 };
 
-struct ComputeDuDt {
+struct ComputeVolumeDuDt {
```
Comment (wthrowe) [inline tests/Unit/Evolution/Actions/Test_ComputeTimeDerivative.cpp]: Why this change?

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp (hunk 15 above)
Comment (nilsvu) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp:80]: I think it's a fine name for this local variable. Didn't want to make it too abstract.

File: tests/Unit/Evolution/Actions/Test_ComputeTimeDerivative.cpp (hunk 16 above)
Comment (nilsvu) [reply tests/Unit/Evolution/Actions/Test_ComputeTimeDerivative.cpp]: left-over from the name changes. reverted

**Comment by nilsvu** [comment]
@wthrowe @nilsdeppe @kidder squashed and rebased

**Review by nilsdeppe — APPROVED** [review]
Restarted GCC debug builds

**Review by wthrowe — APPROVED** [review]

**Review by kidder — APPROVED** [review]
